Rough V1 draft of spec

yondonfu opened the pull request:
This is an initial rough draft of technical spec. I tried to focus on the structure and listing the various protocol actions that we would need to document more than the actual content itself with the goal of filling in more content once we decide how we like this proposed structure.

dob commented:
Overall this looks great and I think is the right level of detail prior to getting feedback from an auditor. I may have missed it, but I didn't see a description of the logic used to update inflation rate or to distribute fees/reward pool proportionally. Perhaps that's ok though if it's really the code that defines it.

dob commented on SPEC.md:
```diff
@@ -0,0 +1,675 @@
+and the Ethereum blockchain. Consequently, this specification assumes smart contracts as the implementation tool and will describe protocol actions in terms of smart contract function calls that
+trigger state transitions within the system.
+
+## Design
```
I think just prior to the design section we should have a 1 or 2 paragraph overview of the progression of the protocol. I can try to write this. I think it will make it easier for someone to absorb the following technical concepts.

dob commented on SPEC.md:
```diff
@@ -0,0 +1,675 @@
+
+Non-attributable faults include:
+
+- **Failed verification**: The transcoder failed verification for a challenged segment. This type of fault is non-attributable because the protocol cannot differentiate from the scenario in which
```
Is a failed verification non-attributable? I could see if verification doesn't return, or returns a known error value that potentially being non-attributable, but if the oracle sends back a different result?

yondonfu replied:
Right - I was thinking about the scenario in which a transcoder fails verification due to non-deterministic transcoding results, but if part of the protocol participation expectations is that transcoders run their nodes using specifically defined architecture, then the failed verification fault can be considered attributable in that context.

dob commented on SPEC.md:
```diff
@@ -0,0 +1,675 @@
+
+Let `jobs` be a mapping of job IDs to transcode jobs. The transcode receipt hash is `keccak256(jobs[jobId].streamId, segmentNumber, dataHashes[0], dataHashes[1], broadcasterSig)`.
+
+#### Transcoder Pool
```
This section might be just a little too specific in the implementation details. It could be simplified to one or two sentences probably.

dob commented on SPEC.md:
```diff
@@ -0,0 +1,675 @@
+5. Set the maximum number of LPT that can be issued in the round based on the current LPT supply and the current `inflationRate`.
+6. Set the maximum number of LPT that can be redistributed for the round based on the current redistribution pool and some constant number of rounds to spread the redistribution across.
+
+#### Notes
```
One idea for the client here is that perhaps the client could have logic to randomly call initialize round with a certain probability if you believe you're in the transcoder set. And then it could sleep for awhile and re-attempt if it wasn't yet initialized by someone. Just an idea in general, not necessarily for the spec.

yondonfu commented:
- Added a Mechanics section that provides a high level overview of protocol progression
- Added a protocol action section on inflation adjustment
- Added a protocol action section on how rewards/fees are paid out to delegators which references the protocol action sections on manually vs. automatically claiming rewards/fees as a delegator
- Added divider lines to more clearly differentiate between the various protocol action descriptions and italicized the sub-section headers within protocol action descriptions
- Updated the Faults section to categorize faults as discoverable by watchers vs. discoverable by an external verification system

yondonfu commented:
Right, it might be a good idea to add some clarification on the percentages in a separate PR
